### PR TITLE
Add cheatsheet for common JS functions

### PR DESCRIPTION
**Pull request by Tawishi**
Solves #1189

I have added the Cheatsheet.md file and linked it with the README.md file.
Sorry for the delay and thank you for your patience.

Please review and suggest changes, if necessary.

### CODE REVIEW

File: Cheatsheet.md [hunk 1]
```diff
@@ -0,0 +1,554 @@
+# Cheatsheet
+### Commonly used [array](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays) functions
```
Comment (Kundan28) [inline Cheatsheet.md]: Rephrase it to `Commonly used functions on Array` and make it header 2 instead of header 3.

File: Cheatsheet.md [hunk 2]
```diff
@@ -0,0 +1,554 @@
+# Cheatsheet
+### Commonly used [array](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays) functions
+
+## [1. map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)
```
Comment (Kundan28) [inline Cheatsheet.md]: make all the functions to header 3 instead of header 2.

File: Cheatsheet.md [hunk 3]
```diff
@@ -0,0 +1,554 @@
+# Cheatsheet
+### Commonly used [array](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays) functions
+
+## [1. map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) 
+
+**Description:**
+creates an array by calling a specific function on each element present in the parent array.
```
Comment (Kundan28) [inline Cheatsheet.md]: creates -> creates

File: Cheatsheet.md [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+
+## [2. filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)
+
+**Description:**
+Creates an array filled with all array elements that pass a test (provided as a function).
+
+**Syntax:**
+ * _array.filter(function(currentValue, index, arr), thisValue)_
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 28 or 29.

File: Cheatsheet.md [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+**Return value:**
+Array containing all elements that pass the test else an empty array.
+
+**Example:**
+
+    function isPositive(value) {
+      return value > 0;
+    }
+    var filtered = [112, 52, 0, -1, 944].filter(isPositive);
+    print(filtered);
```
Comment (Kundan28) [inline Cheatsheet.md]: Enclose it under **```**
It's not single quote, it's called back tick.
You have to write that before and after the code.

File: Cheatsheet.md [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+| function(currentValue,index,arr) | Required function to be run for each array element |
+|initialValue | Optional.|
+
+**Return value:**
+returns accumulated result fron the last call of the callback function
+
+**Example:**
+
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 67 & 68

File: Cheatsheet.md [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+ **Return value:**
+ New array containing the selected elements.
+
+ **Example:**
+
+    var arr = [23,56,87,32,75,13];
+    var new_arr = arr.slice();
+    print(arr);
+    print(new_arr);
```
Comment (Kundan28) [inline Cheatsheet.md]: Enclose the code in back ticks.

File: Cheatsheet.md [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+*Output:*
+
+    [23,56,87,32,75,13]
+    [23,56,87,32,75,13]
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 96

File: Cheatsheet.md [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+Array without the removed items.
+
+**Example:**
+
+    var languages = ['C++', 'Java', 'Html', 'Python', 'C']; 
+    document.write(languages + "<br>"); 
+    // Add 'Julia' and 'Php' after removing 'Html'. 
+    var removed = languages.splice(2, 1, 'Julia', 'Php') 
+    document.write(languages + "<br>"); 
+    document.write(removed + "<br>");
```
Comment (Kundan28) [inline Cheatsheet.md]: Enclose the code in back ticks.

File: Cheatsheet.md [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+
+*Output:*
+
+    C++,Java,Html,Python,C
+    C++,Java,Julia,Php,Python,C
+
+					 
+ 
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 128-131

File: Cheatsheet.md [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+**Return value:**
+joined array
+
+**Example:**
+
+    var num1 = [11, 12, 13],
+    num2 = [14, 15, 16],
+    num3 = [17, 18, 19];
+    print(num1.concat(num2, num3));
```
Comment (Kundan28) [inline Cheatsheet.md]: Enclose this in back ticks.

File: Cheatsheet.md [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+
+*Output:*
+
+    [11,12,13,14,15,16,17,18,19]
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 158

File: Cheatsheet.md [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+| separator	| Optional.   | 
+
+**Return value:**
+String, representing the array values, separated by the specified separator.
+
+**Example:**
+
+    var elements = [\'geeksforgeeks\', \'gfg\']; console.log(elements.join()); 
+    console.log(elements.join(\'.\')); 
+    console.log(elements.join(\'-\'));
```
Comment (Kundan28) [inline Cheatsheet.md]: Enclose this under back ticks.

File: Cheatsheet.md [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+    var elements = [\'geeksforgeeks\', \'gfg\']; console.log(elements.join()); 
+    console.log(elements.join(\'.\')); 
+    console.log(elements.join(\'-\')); 
+*Output:*
+
+    > \"geeksforgeeks, gfg\"
+    > \"geeksforgeeks.gfg\"
+    > \"geeksforgeeks-gfg\"
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: remove line 186

File: Cheatsheet.md [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+    // Printing desired values. 
+      console.log(found); 
+*Output:*
+    
+    > 7
+
+
+
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 241-244

File: Cheatsheet.md [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+**Example:**
+
+    print('Departed Train'.indexOf('Train'));
+*Output:*
+    
+    9
+
+---
+---
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 267-270

File: Cheatsheet.md [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+
+
+
+
+
+
+
+
+
```
Comment (Kundan28) [inline Cheatsheet.md]: Remove line 521-554

File: README.md [hunk 18]
```diff
@@ -35,6 +35,8 @@ The following is the approximate sequence of the topics during the program but i
 
 ### Week 0 - Pre-Work
 Before beginning the program, apprentices meet weekly for 4 weeks to learn JavaScript fundamentals.
+Apprentices should memorize [**this**](Cheatsheet.md) within the first month.
+
```
Comment (Kundan28) [inline README.md]: Remove extra line space

File: Cheatsheet.md (hunk 2 above)
Comment (Tawishi) [reply Cheatsheet.md]: When I change it to header 3 the headings disappear in the preview.

Comment (Kundan28) [reply Cheatsheet.md]: Make it change as it doesn't look good that the main header is smaller that the sub header.

Comment (Tawishi) [reply Cheatsheet.md]: Would same size be OK ?

Comment (Kundan28) [reply Cheatsheet.md]: Same size would be little bit off.
Let's ask @alodahl regarding this.

Comment (Tawishi) [reply Cheatsheet.md]: Done.

File: Cheatsheet.md (hunk 3 above)
Comment (Tawishi) [reply Cheatsheet.md]: Did you mean 
create -> Create ?

File: Cheatsheet.md [hunk 19]
```diff
@@ -259,19 +227,18 @@ returns the position of the first occurrence of a specified value in a string.
...
-
+```
     print('Departed Train'.indexOf('Train'));
+```
 *Output:*
     
     9
 
 ---
 ---
```
Comment (Kundan28) [inline Cheatsheet.md]: Are these lines required ?

File: Cheatsheet.md [hunk 20]
```diff
@@ -387,21 +347,17 @@ Returns the character at the specified index in a string.
...
     print(str.charAt(9));
-  
+```  
   *Output:*
 
     t
-
-
 ---
 ---
```
Comment (Kundan28) [inline Cheatsheet.md]: Same here

File: Cheatsheet.md (hunk 19 above)
Comment (Tawishi) [reply Cheatsheet.md]: These are two horizontal lines separating the header contents. Earlier the size was smaller for the headings that is why I had inerted these. Should I remove them now?

Comment (Kundan28) [reply Cheatsheet.md]: Yeah, remove them.
And I said to remove extra line space, you remove all of them.
Please fix that too.

Comment (Tawishi) [reply Cheatsheet.md]: Ok.

**Comment by Kundan28** [comment]
Extra line space means if there are 2 line space remove one from there not both of them.

**Comment by Tawishi** [comment]
> Extra line space means if there are 2 line space remove one from there not both of them.

Ok, sir.

**Comment by Tawishi** [comment]
> @Tawishi Revert back this change, I said remove extra line space and you are removing the required ones.

Which changes?

**Comment by Kundan28** [comment]
> Which changes?

That you made in README.md file and the similar mistake you did in other file.

**Comment by Tawishi** [comment]
> You have to same thing in entire file. And revert the changes in README.md

Please check if the same problem persists.

**Comment by Kundan28** [comment]
You have not made the changes in the entire file which I reviewed the first

**Comment by Tawishi** [comment]
> You have not made the changes in the entire file which I reviewed the first
Sir, if I may ask , the preview is not affected by these changes then why are these necessary?

File: Cheatsheet.md [hunk 21]
```diff
@@ -0,0 +1,504 @@
+# Cheatsheet
+## Commonly used functions on [Array](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays)
+### [**1. map**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)
```
Comment (alodahl) [inline Cheatsheet.md]: please follow proper md formatting.  These ### headers should not be bolded as well, please remove the ** from all of them.

**Comment by Kundan28** [comment]
> Sir, if I may ask , the preview is not affected by these changes then why are these necessary?

If the preview is same, then why to make the changes in first place.

File: Cheatsheet.md (hunk 21 above)
Comment (Tawishi) [reply Cheatsheet.md]: ok.

Comment (Tawishi) [reply Cheatsheet.md]: > ok.

done

**Comment by Tawishi** [comment]
> Nice work @Tawishi
Thank you, sir. Wouldn't have been possible without your guidance.

File: Cheatsheet.md [hunk 22]
```diff
@@ -0,0 +1,504 @@
...
+
+**Syntax:**
+ * _array.push(item1, item2, ..., itemX)_
+
+| Parameter | Description |
+| :-------- | :---------- |
+| item1, item2, ..., itemX | Required. The item(s) to add to the array |
+
+**Return value:**
+New length of hte array.
```
Comment (alodahl) [inline Cheatsheet.md]:
```suggestion
New length of the array.

```

File: Cheatsheet.md [hunk 23]
```diff
@@ -0,0 +1,504 @@
...
+ * _array.reduce( function(total, currentValue, currentIndex, arr), initialValue )_
+
+ | Parameter | Description |
+| :-------- | :---------- |
+| function(currentValue,index,arr) | Required function to be run for each array element |
+|initialValue | Optional.|
+
+**Return value:**
+returns accumulated result fron the last call of the callback function
+
```
Comment (alodahl) [inline Cheatsheet.md]: All of these examples are so helpful!  Can `reduce` have an example too?

File: Cheatsheet.md [hunk 24]
```diff
@@ -0,0 +1,504 @@
...
+**Syntax:**
+ * _array.find(function(currentValue, index, arr),thisValue)_
+
+| Parameter | Description |
+| :-------- | :---------- |
+| function(currentValue, index, arr) |	Required. A function to be run for each element in the array.|
+| thisValue	| Optional. |
+
+**Return value:**
+ The array element value if any of the elements in the array pass the test, otherwise it returns undefine.
```
Comment (alodahl) [inline Cheatsheet.md]:
```suggestion
 The array element value if any of the elements in the array pass the test, otherwise it returns `undefined`.

```

File: Cheatsheet.md [hunk 25]
```diff
@@ -0,0 +1,504 @@
...
+
+**Example:**
+ ```  
+    var array = [2, 7, 8, 9]; 
+    var found = array.find(function(element) {
```
Comment (alodahl) [inline Cheatsheet.md]:
```suggestion
    const found = array.find(element => element > 4);

```

File: Cheatsheet.md [hunk 26]
```diff
@@ -0,0 +1,504 @@
...
+| thisValue	| Optional. |
+
+**Return value:**
+ The array element value if any of the elements in the array pass the test, otherwise it returns undefine.
+
+**Example:**
+ ```  
+    var array = [2, 7, 8, 9]; 
+    var found = array.find(function(element) { 
+    return element > 4;
```
Comment (alodahl) [inline Cheatsheet.md]:
```suggestion
```

File: Cheatsheet.md [hunk 27]
```diff
@@ -0,0 +1,504 @@
...
+    // Taking a string as variable 
+    var string = "geeksforgeeks"; 
+    a = string.substring(0, 4) 
+    b = string.substring(1, 6) 
+    c = string.substring(5) 
+    d = string.substring(0) 
+  
+    // Printing new string which are 
+    // the part of the given string 
+    document.write(a + "<br>");
```
Comment (alodahl) [inline Cheatsheet.md]: please use console.log instead.

**Comment by alodahl** [comment]
I moved the `solves #` statement to the description and the PR title now reflects the changes made in the PR.  Please title and describe your PRs this way in future issues.

File: Cheatsheet.md [hunk 28]
```diff
@@ -0,0 +1,504 @@
+# Cheatsheet
+## Commonly used functions on [Array](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays)
+### [1. map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)
+
+**Description:**
+Creates an array by calling a specific function on each element present in the parent array.
+
+**Syntax:**
+* _array.map(function(currentValue, index, arr), thisValue)_
```
Comment (vegetabill) [inline Cheatsheet.md]: Only include the first parameter, currentValue instead of the others which are rarely used.

The same goes for all callbacks in this doc. Only include the first argument.

File: Cheatsheet.md [hunk 29]
```diff
@@ -0,0 +1,504 @@
+# Cheatsheet
+## Commonly used functions on [Array](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays)
+### [1. map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)
```
Comment (vegetabill) [inline Cheatsheet.md]: Please remove the numbering from these - it will make the hashlinks break more easily

File: Cheatsheet.md [hunk 30]
```diff
@@ -0,0 +1,504 @@
...
+**Return value:** 
+An Array containing the results of calling the provided function for each element in the original array.
+
+### [2. filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)
+
+**Description:**
+Creates an array filled with all array elements that pass a test (provided as a function).
+
+**Syntax:**
+ * _array.filter(function(currentValue, index, arr), thisValue)_
```
Comment (vegetabill) [inline Cheatsheet.md]: All these should be surrounded by backticks, not italics

File: Cheatsheet.md [hunk 31]
```diff
@@ -0,0 +1,504 @@
...
+| thisValue | Optional. |
+
+**Return value:**
+Array containing all elements that pass the test else an empty array.
+
+**Example:**
+```
+let filtered = [112, 52, 0, -1, 944].filter(value => value>0);
+console.log(filtered);
+ ```
```
Comment (vegetabill) [inline Cheatsheet.md]: all example code should be noted as JavaScript for proper highlighting

see  https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code

File: Cheatsheet.md [hunk 32]
```diff
@@ -0,0 +1,504 @@
+# Cheatsheet
+## Commonly used functions on [Array](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays)
+### [1. map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)
+
+**Description:**
```
Comment (vegetabill) [inline Cheatsheet.md]: These should all be 4th level headings, not just bold (remove the colon too)

File: Cheatsheet.md [hunk 33]
```diff
@@ -0,0 +1,504 @@
...
+| :-------- | :---------- |
+| function(currentValue,index,arr) | Required function to be run for each array element |
+| thisValue | Optional. |
+
+**Return value:**
+Array containing all elements that pass the test else an empty array.
+
+**Example:**
+```
+let filtered = [112, 52, 0, -1, 944].filter(value => value>0);
```
Comment (vegetabill) [inline Cheatsheet.md]: please add spaces like `value > 0`

and for **all example code** in this doc, please use `function` functions, not `=>` arrow functions since this doc is for beginners

File: Cheatsheet.md [hunk 34]
```diff
@@ -0,0 +1,504 @@
...
+```
+    let arr = [23,56,87,32,75,13];
+    let new_arr = arr.slice();
+    console.log(arr);
+    console.log(new_arr);
+```    
+*Output:*
+
+    [23,56,87,32,75,13]
+    [23,56,87,32,75,13]
```
Comment (vegetabill) [inline Cheatsheet.md]: please use an example that actually takes a slice

File: Cheatsheet.md [hunk 35]
```diff
@@ -0,0 +1,504 @@
...
+
+ **Return value:**
+ New array containing the selected elements.
+
+ **Example:**
+```
+    let arr = [23,56,87,32,75,13];
+    let new_arr = arr.slice();
+    console.log(arr);
+    console.log(new_arr);
```
Comment (vegetabill) [inline Cheatsheet.md]: do not put console.log in examples. it's visual noise

File: Cheatsheet.md [hunk 36]
```diff
@@ -0,0 +1,504 @@
...
+```    
+*Output:*
+
+    [23,56,87,32,75,13]
+    [23,56,87,32,75,13]
+
+### [5. splice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice)
+
+**Description:**
+Adds/removes items to/from an array, and returns the removed item(s).
```
Comment (vegetabill) [inline Cheatsheet.md]: add a note about the difference between `slice` and `splice`. it is a common point of confusion among JavaScript learners

File: Cheatsheet.md [hunk 37]
```diff
@@ -0,0 +1,504 @@
...
+console.log(array1.reduce(reducer, 5));
+```
+*Output:*
+    
+    10
+    15
+
+### [4. slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice)
+
+**Description:**
```
Comment (vegetabill) [inline Cheatsheet.md]: add a note about the difference between `slice` and `splice`. it is a common point of confusion among JavaScript learners

File: Cheatsheet.md [hunk 38]
```diff
@@ -0,0 +1,504 @@
...
+| Parameter | Description |
+| :-------- | :---------- |
+| separator	| Optional.   | 
+
+**Return value:**
+String, representing the array values, separated by the specified separator.
+
+**Example:**
+```
+const elements = [\'geeksforgeeks\', \'gfg\']; console.log(elements.join());
```
Comment (vegetabill) [inline Cheatsheet.md]: Please remove the escaping on the strings. Also please use the same data from above, the list of languages.
Again, do not use console.log and just list the output

File: Cheatsheet.md [hunk 39]
```diff
@@ -0,0 +1,504 @@
...
+**Example:**
+```
+let arr = [34, 234, 567, 4];
+console.log(arr.push(23,45,56));
+console.log(arr);
+```
+*Output:*
+
+    7
+    34,234,567,4,23,45,56
```
Comment (vegetabill) [inline Cheatsheet.md]: please use smaller numbers like `[1, 2, 3, 4]` which is easier to read
(and like all others, remove console.log)

File: Cheatsheet.md [hunk 40]
```diff
@@ -0,0 +1,504 @@
...
+```
+*Output:*
+
+    C++,Java,Html,Python,C
+    C++,Java,Julia,Php,Python,C
+
+### [6. concat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat)
+
+**Description:**
+used to join two or more arrays.
```
Comment (vegetabill) [inline Cheatsheet.md]: It can also be used to add a single item (not an array) and that is how it is more commonly used so please focus your description and examples on that.

Please add a note comparing it to `.push` in that this does not mutate the array but creates a new one.

File: Cheatsheet.md [hunk 41]
```diff
@@ -0,0 +1,504 @@
...
+
+**Example:**
+```
+console.log('Departed Train'.indexOf('Train'));
+```
+*Output:*
+    
+    9
+    
+## Commonly used functions on [string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)
```
Comment (vegetabill) [inline Cheatsheet.md]: Please make the heading just "Common String Functions", like this:

(same for Array as well)

```markdown
## Common String Functions
[Full Mozilla MDN Docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)
```

File: Cheatsheet.md [hunk 42]
```diff
@@ -0,0 +1,504 @@
...
+*Output:*
+    
+    9
+    
+## Commonly used functions on [string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) 
+
+### [1. substring](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/substring)
+
+**Description:**
+Extracts the characters in a string between "start" and "end", not including "end" itself.
```
Comment (vegetabill) [inline Cheatsheet.md]: Add a note with the similarities/differences between this and array.slice since beginners get confused on them

File: Cheatsheet.md [hunk 43]
```diff
@@ -0,0 +1,504 @@
...
+|start	| Required. The position where to start the extraction. First character is at index 0 |
+| end	| Optional. The position (up to, but not including) where to end the extraction. If omitted, it extracts the rest of the string|
+
+**Return value:**
+New string containing the extracted characters.
+
+**Example:**
+```
+// Taking a string as variable 
+const string = "geeksforgeeks";
```
Comment (vegetabill) [inline Cheatsheet.md]: good example 👍

File: Cheatsheet.md [hunk 44]
```diff
@@ -0,0 +1,504 @@
...
+| :-------- | :---------- |
+|separator	| Optional. Specifies the character, or the regular expression, to use for splitting the string. If omitted, the entire string will be returned (an array with only one item)|
+| limit	| Optional. An integer that specifies the number of splits, items after the split limit will not be included in the array |
+
+**Return value:**
+Array containing splitted values.
+
+**Example:**
+```
+let str = 'It iS a 5r&e@@t Day.'
```
Comment (vegetabill) [inline Cheatsheet.md]: please use different test data, as leet-speak isn't comfortable for beginners if they don't know about it

File: Cheatsheet.md [hunk 45]
```diff
@@ -0,0 +1,504 @@
...
+let array = str.split(" ");
+console.log(array);
+```
+*Output:*
+
+    [It,iS,a,5r&e@@t,Day.]
```
Comment (vegetabill) [inline Cheatsheet.md]: Actually this is _not_ the output. The values should be quoted

File: Cheatsheet.md [hunk 46]
```diff
@@ -0,0 +1,504 @@
...
+
+**Example:**
+```
+const str = "GeeksforGeeks      ";
+const st = str.trim();
+console.log(st);
+```
+*Output:*
+   
+    GeeksForGeeks
```
Comment (vegetabill) [inline Cheatsheet.md]: add quotes to make it clear what was trimmed

File: Cheatsheet.md [hunk 47]
```diff
@@ -0,0 +1,504 @@
...
+
+**Syntax:**
+ * _string.trim()_
+
+**Return value:**
+String, representing the string with removed whitespace from both ends
+
+**Example:**
+```
+const str = "GeeksforGeeks      ";
```
Comment (vegetabill) [inline Cheatsheet.md]: add some leading whitespace too

File: Cheatsheet.md [hunk 48]
```diff
@@ -0,0 +1,504 @@
...
+**Example:**
+```
+const str = 'JavaScript is object oriented language';
+console.log(str.charAt(9));
+```  
+  *Output:*
+
+    t
+    
+## Many forms of for loops
```
Comment (vegetabill) [inline Cheatsheet.md]: Make this heading into "For Loop Variants"

File: Cheatsheet.md [hunk 49]
```diff
@@ -0,0 +1,504 @@
...
+
+### [1. for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in)
+
+**Description:**
+ Loops through the properties of an object. The block of code inside the loop will be executed once for each property.
+
+**Syntax:**
+ * _for (var in object) {<br>
+  code block to be executed<br>
+ }_
```
Comment (vegetabill) [inline Cheatsheet.md]: The example shows the syntax, you don't need another one

File: Cheatsheet.md [hunk 50]
```diff
@@ -0,0 +1,504 @@
...
+
+| Parameter | Description |
+| :---      | :---        |
+|var	|Required. A variable that iterates over the properties of an object|
+|object	| Required. The specified object that will be iterated|
+
```
Comment (vegetabill) [inline Cheatsheet.md]: These are language constructs, not functions. This kind of doc isn't necessary. Please remove this

File: Cheatsheet.md [hunk 51]
```diff
@@ -0,0 +1,504 @@
...
+  *Output:*
+
+    t
+    
+## Many forms of for loops
+
+### [1. for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in)
+
+**Description:**
+ Loops through the properties of an object. The block of code inside the loop will be executed once for each property.
```
Comment (vegetabill) [inline Cheatsheet.md]: These are language constructs, not functions. This kind of doc isn't necessary. Please remove this

File: Cheatsheet.md [hunk 52]
```diff
@@ -0,0 +1,504 @@
...
+*Output:*
+
+    "a"
+    "b"
+    "c"
+
+### [2.for...of](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/for...of)
+
+**Description:**
+Loops through the values of an iterable objects.
```
Comment (vegetabill) [inline Cheatsheet.md]: Beginner will not know what "iterable" means. Please be more concrete

File: Cheatsheet.md [hunk 53]
```diff
@@ -0,0 +1,504 @@
...
+
+**Syntax:**
+ * _for (variable of iterable) {<br>
+  // code block to be executed<br>
+}_
+
+| Parameter | Description |
+| :---      | :---        |
+| variable | For every iteration the value of the next property is assigned to the variable. Variable can be declared with const, let, or var.|
+|iterable | An object that has iterable properties.|
```
Comment (vegetabill) [inline Cheatsheet.md]: These are language constructs, not functions. This kind of doc isn't necessary. Please remove this

File: Cheatsheet.md [hunk 54]
```diff
@@ -0,0 +1,504 @@
...
+```  
+const languages = { first : "C", second : "Java", 
+			third : "Python", fourth : "PHP", 
+		        	fifth : "JavaScript" }; 
+// iterate through every property of the 
+// object languages and print all of them 
+// using for..in loops 
+for (itr in languages) 
+{ 
+	console.log(languages[itr]);
```
Comment (vegetabill) [inline Cheatsheet.md]: console.log is a good choice here as a demo of the loop

File: Cheatsheet.md [hunk 55]
```diff
@@ -0,0 +1,504 @@
...
+
+**Example:**
+```  
+const languages = { first : "C", second : "Java", 
+			third : "Python", fourth : "PHP", 
+		        	fifth : "JavaScript" }; 
+// iterate through every property of the 
+// object languages and print all of them 
+// using for..in loops 
+for (itr in languages)
```
Comment (vegetabill) [inline Cheatsheet.md]: This example doesn't match the section, as it's using `in`

File: Cheatsheet.md [hunk 56]
```diff
@@ -0,0 +1,504 @@
...
+**Syntax:**
+ * _for (statement 1; statement 2; statement 3) {<br>
+  // code block to be executed<br>
+}_
+
+| Parameter | Description |
+| :---      | :---        |
+| _statement 1_ : variable declaration| Executed (one time) before the execution of the code block.|
+| _statement 2_ : test condition| Defines the condition for executing the code block. |
+| _statement 3_ : change in variable| Executed (every time) after the code block has been executed. |
```
Comment (vegetabill) [inline Cheatsheet.md]: Remove all this and replace with a student-friendly description: something like "traditional index-based loop as found in many languages"

File: Cheatsheet.md (hunk 28 above)
Comment (Tawishi) [reply Cheatsheet.md]: Sir, are you referring to parameters of _function_ or the parameters of _array.map( )_?

File: Cheatsheet.md (hunk 34 above)
Comment (Tawishi) [reply Cheatsheet.md]: Ok, sir.

File: Cheatsheet.md (hunk 35 above)
Comment (Tawishi) [reply Cheatsheet.md]: Sir, I am unable to understand. Do you mean to say that there is no need to print the output?

File: Cheatsheet.md (hunk 36 above)
Comment (Tawishi) [reply Cheatsheet.md]: Ok, sir.

Comment (Tawishi) [reply Cheatsheet.md]: Should I also include split in this comparison?

File: Cheatsheet.md (hunk 29 above)
Comment (Tawishi) [reply Cheatsheet.md]: Done.

File: Cheatsheet.md (hunk 31 above)
Comment (Tawishi) [reply Cheatsheet.md]: Done.

File: Cheatsheet.md (hunk 30 above)
Comment (Tawishi) [reply Cheatsheet.md]: Done. Should we keep the bullets?

File: Cheatsheet.md (hunk 32 above)
Comment (Tawishi) [reply Cheatsheet.md]: Bold is not working with fourth level headings. So, they are now unbolded fourth level headings.

File: Cheatsheet.md (hunk 33 above)
Comment (Tawishi) [reply Cheatsheet.md]: Done.

File: Cheatsheet.md (hunk 38 above)
Comment (Tawishi) [reply Cheatsheet.md]: Got it.

File: Cheatsheet.md (hunk 39 above)
Comment (Tawishi) [reply Cheatsheet.md]: Got it.

File: Cheatsheet.md (hunk 56 above)
Comment (Tawishi) [reply Cheatsheet.md]: Replace this description or all the descriptions?

File: Cheatsheet.md (hunk 53 above)
Comment (Tawishi) [reply Cheatsheet.md]: Sir, are you referring to the table?

File: Cheatsheet.md (hunk 51 above)
Comment (Tawishi) [reply Cheatsheet.md]: No description?

Comment (Tawishi) [reply Cheatsheet.md]: Only for this function or for all the functions?

File: Cheatsheet.md (hunk 49 above)
Comment (Tawishi) [reply Cheatsheet.md]: Only this example?

File: Cheatsheet.md (hunk 45 above)
Comment (Tawishi) [reply Cheatsheet.md]: Done.

File: Cheatsheet.md (hunk 41 above)
Comment (Tawishi) [reply Cheatsheet.md]: Done.

File: Cheatsheet.md (hunk 35 above)
Comment (Tawishi) [reply Cheatsheet.md]: Got it.

**Comment by Tawishi** [comment]
Please review for any more changes, if needed @Kundan28, @alodahl and @vegetabill.

File: Cheatsheet.md (hunk 56 above)
Comment (alodahl) [reply Cheatsheet.md]: @vegetabill ? ^

**Comment by Kundan28** [comment]
Updated in scoresheet.

**Comment by Tawishi** [comment]
Thank you for your time, guidance and patience @Kundan28, @alodahl and @vegetabill.

**Comment by Kundan28** [comment]
> Thank you for your time, guidance and patience @Kundan28, @alodahl and @vegetabill.

No problem.
